Fixed errors based on cases in dump from server

bansalnitish opened the pull request:
This PR updates the TextAngular migration process to fix the errors based on cases obtained in dump from server.

bansalnitish committed:
Updated text angular migration according to errors from dump

bansalnitish committed:
Updated error yielding

seanlip commented on core/domain/html_cleaner.py:
```diff
@@ -182,7 +182,13 @@ def unescape_html(escaped_html_data):
+    # Any such extra quotes are not rendered as \&quot; in html
+    # string. They are rendered in form of &amp;quot;. This line
+    # ensures that all such extra quotes are removed from the html string.
+    unescaped_html_data = escaped_html_data.replace('&amp;quot;', '')
```
How do you know that this does not replace more stuff than you want it to? Can we have a (much) more restrictive check?

seanlip commented on core/domain/html_cleaner.py:
```diff
@@ -573,9 +579,21 @@ def convert_tag_contents_to_rte_format(html_data, rte_conversion_fn):
+            collapsible['content-with-value'] = escape_html(json.dumps(''))
+        else:
+            content_html = unescape_html(collapsible['content-with-value'])
+            converted_html = ''
```
```
converted_html = (
    rte_conversion_fn(json.loads(content_html)) if content_html else '')
```

(and put the comment below above this line)

seanlip commented on core/domain/html_cleaner.py:
```diff
@@ -632,12 +650,16 @@ def validate_rte_format(html_list, rte_format, run_migration=False):
+                is_invalid = True
+            else:
+                content_html = unescape_html(collapsible['content-with-value'])
+                if len(content_html):
```
Wait -- at this point, why do we need this check? Shouldn't all content_html be json-parseable by this point?

bansalnitish replied:
When the `content-with-value=""` then json.loads() gives an error 
```
("No JSON object could be decoded")\nValueError: No JSON object could be decoded\n'
```

seanlip replied:
Why is content-with-value = "" possible at this point?  Shouldn't we be catching that and fixing it rather than skipping over it?

bansalnitish replied:
OK , I got it now. I have made the fix for this, PTAL!

AllanYangZhou commented on core/domain/html_cleaner.py:
```diff
@@ -182,7 +182,13 @@ def unescape_html(escaped_html_data):
+    # include double quotes around the font name. This results
+    # in error while performing json.loads.
+    # Any such extra quotes are not rendered as \&quot; in html
+    # string. They are rendered in form of &amp;quot;. This line
```
Interesting, what does this content look like in the frontend? The frontend has to do a similar json conversion process, so it seems like that would error due to the extra double quotes.

What will your change  (removing the extra double quotes) do, visually? Won't simply removing the double quotes make the style tag invalid?

bansalnitish replied:
I have given a example below, PTAL.

AllanYangZhou replied:
The comment can be simplified:

Some content erroneously contains un-escaped double quotes (`&amp;quot;` instead of `\&amp;quot;`) which breaks json parsing. This line ensures all double quotes are escaped.

Edit: Aside from this I have no other issues, once @seanlip's concerns are addressed we can merge.

bansalnitish commented:
Hi @seanlip @AllanYangZhou Regarding the issue with quotes within style properties:
The case is as follows:
```
<oppia-noninteractive-collapsible content-with-value="&amp;quot;&amp;lt;span style=\\&amp;quot;font-size:12.0pt;font-family:&amp;amp;quot;Times New Roman&amp;amp;quot;;\\nmso-fareast-font-family:&amp;amp;quot;Times New Roman&amp;amp;quot;;mso-font-kerning:12.0pt;mso-ansi-language:\\nHR;mso-fareast-language:HR;mso-bidi-language:AR-SA;mso-bidi-font-weight:bold\\&amp;quot;&amp;gt;Kao\\nprimjer: heksadecimalni broj&amp;amp;nbsp;&amp;lt;/span&amp;gt;&amp;lt;span style=\\&amp;quot;font-size: 14px;\\&amp;quot;&amp;gt;A&amp;lt;/span&amp;gt;&amp;lt;sub&amp;gt;(16)&amp;lt;/sub&amp;gt;&amp;lt;span style=\\&amp;quot;font-family: &amp;#39;Times New Roman&amp;#39;; font-size: 12pt;\\&amp;quot;&amp;gt;&amp;amp;nbsp;mo\u017eemo pretvoriti u binarni na sljede\u0107i\\nna\u010din:&amp;lt;/span&amp;gt;&amp;lt;span style=\\&amp;quot;font-size:12.0pt;font-family:&amp;amp;quot;Times New Roman&amp;amp;quot;;\\nmso-fareast-font-family:&amp;amp;quot;Times New Roman&amp;amp;quot;;mso-font-kerning:12.0pt;mso-ansi-language:\\nHR;mso-fareast-language:HR;mso-bidi-language:AR-SA;mso-bidi-font-weight:bold\\&amp;quot;&amp;gt;&amp;lt;br&amp;gt;&amp;lt;/span&amp;gt;&amp;lt;div&amp;gt;&amp;lt;span style=\\&amp;quot;font-size:12.0pt;font-family:&amp;amp;quot;Times New Roman&amp;amp;quot;;\\nmso-fareast-font-family:&amp;amp;quot;Times New Roman&amp;amp;quot;;mso-font-kerning:12.0pt;mso-ansi-language:\\nHR;mso-fareast-language:HR;mso-bidi-language:AR-SA;mso-bidi-font-weight:bold\\&amp;quot;&amp;gt;&amp;lt;br&amp;gt;&amp;lt;/span&amp;gt;&amp;lt;/div&amp;gt;&amp;lt;div&amp;gt;&amp;lt;oppia-noninteractive-image filepath-with-value=\\&amp;quot;&amp;amp;amp;quot;xxxxxxxxxxxxxx.jpeg&amp;amp;amp;quot;\\&amp;quot;&amp;gt;&amp;lt;/oppia-noninteractive-image&amp;gt;&amp;lt;span style=\\&amp;quot;font-size:12.0pt;font-family:&amp;amp;quot;Times New Roman&amp;amp;quot;;\\nmso-fareast-font-family:&amp;amp;quot;Times New Roman&amp;amp;quot;;mso-font-kerning:12.0pt;mso-ansi-language:\\nHR;mso-fareast-language:HR;mso-bidi-language:AR-SA;mso-bidi-font-weight:bold\\&amp;quot;&amp;gt;&amp;lt;br&amp;gt;&amp;lt;/span&amp;gt;&amp;lt;/div&amp;gt;&amp;quot;" heading-with-value="&amp;quot;Pomo\u0107&amp;quot;"></oppia-noninteractive-collapsible>
```
The content-with-value before unescaping is as follows:

```
&quot;&lt;span style=\&quot;font-size:12.0pt;font-family:&amp;quot;Times New Roman&amp;quot;;\nmso-fareast-font-family:&amp;quot;Times New Roman&amp;quot;;mso-font-kerning:12.0pt;mso-ansi-language:\nHR;mso-fareast-language:HR;mso-bidi-language:AR-SA;mso-bidi-font-weight:bold\&quot;&gt;Kao\nprimjer: heksadecimalni broj&amp;nbsp;&lt;/span&gt;&lt;span style=\&quot;font-size: 14px;\&quot;&gt;A&lt;/span&gt;&lt;sub&gt;(16)&lt;/sub&gt;&lt;span style=\&quot;font-family: &#39;Times New Roman&#39;; font-size: 12pt;\&quot;&gt;&amp;nbsp;mo\u017eemo pretvoriti u binarni na sljede\u0107i\nna\u010din:&lt;/span&gt;&lt;span style=\&quot;font-size:12.0pt;font-family:&amp;quot;Times New Roman&amp;quot;;\nmso-fareast-font-family:&amp;quot;Times New Roman&amp;quot;;mso-font-kerning:12.0pt;mso-ansi-language:\nHR;mso-fareast-language:HR;mso-bidi-language:AR-SA;mso-bidi-font-weight:bold\&quot;&gt;&lt;br&gt;&lt;/span&gt;&lt;div&gt;&lt;span style=\&quot;font-size:12.0pt;font-family:&amp;quot;Times New Roman&amp;quot;;\nmso-fareast-font-family:&amp;quot;Times New Roman&amp;quot;;mso-font-kerning:12.0pt;mso-ansi-language:\nHR;mso-fareast-language:HR;mso-bidi-language:AR-SA;mso-bidi-font-weight:bold\&quot;&gt;&lt;br&gt;&lt;/span&gt;&lt;/div&gt;&lt;div&gt;&lt;oppia-noninteractive-image filepath-with-value=\&quot;&amp;amp;quot;xxxxxxxxxxxxxx.jpeg&amp;amp;quot;\&quot;&gt;&lt;/oppia-noninteractive-image&gt;&lt;span style=\&quot;font-size:12.0pt;font-family:&amp;quot;Times New Roman&amp;quot;;\nmso-fareast-font-family:&amp;quot;Times New Roman&amp;quot;;mso-font-kerning:12.0pt;mso-ansi-language:\nHR;mso-fareast-language:HR;mso-bidi-language:AR-SA;mso-bidi-font-weight:bold\&quot;&gt;&lt;br&gt;&lt;/span&gt;&lt;/div&gt;&quot;
```
After unescaping the content-with-value is as follows:

```
"<span style=\"font-size:12.0pt;font-family:"Times New Roman";\nmso-fareast-font-family:"Times New Roman";mso-font-kerning:12.0pt;mso-ansi-language:\nHR;mso-fareast-language:HR;mso-bidi-language:AR-SA;mso-bidi-font-weight:bold\">Kao\nprimjer: heksadecimalni broj&nbsp;</span><span style=\"font-size: 14px;\">A</span><sub>(16)</sub><span style=\"font-family: 'Times New Roman'; font-size: 12pt;\">&nbsp;mo\u017eemo pretvoriti u binarni na sljede\u0107i\nna\u010din:</span><span style=\"font-size:12.0pt;font-family:"Times New Roman";\nmso-fareast-font-family:"Times New Roman";mso-font-kerning:12.0pt;mso-ansi-language:\nHR;mso-fareast-language:HR;mso-bidi-language:AR-SA;mso-bidi-font-weight:bold\"><br></span><div><span style=\"font-size:12.0pt;font-family:"Times New Roman";\nmso-fareast-font-family:"Times New Roman";mso-font-kerning:12.0pt;mso-ansi-language:\nHR;mso-fareast-language:HR;mso-bidi-language:AR-SA;mso-bidi-font-weight:bold\"><br></span></div><div><oppia-noninteractive-image filepath-with-value=\"&amp;quot;xxxxxxxxxxxxxx.jpeg&amp;quot;\"></oppia-noninteractive-image><span style=\"font-size:12.0pt;font-family:"Times New Roman";\nmso-fareast-font-family:"Times New Roman";mso-font-kerning:12.0pt;mso-ansi-language:\nHR;mso-fareast-language:HR;mso-bidi-language:AR-SA;mso-bidi-font-weight:bold\"><br></span></div>"
```
The issue is with the double quotes after font-family. The issue is same in other cases as well. I was not able to find a way to make it more restrictive. The only way is that we replace `&amp;quot;` only if it is after font-family attribute in style tags.

What do you think?

bansalnitish commented:
cc / @ishucr7 (Since, he too was facing the above error)

bansalnitish committed:
Minor nit

bansalnitish committed:
Replaced &amp;quot; with \&amp;quot;

bansalnitish commented:
Hi @seanlip @AllanYangZhou 

As per discussion with Allan, he was concerned about the approach I used first to replace `&amp;quot;` with `''`. In this case it can be some double quot is replaced with empty string.

We then decided to replace `&amp;quot;` with `\&amp;quot;`. Also this `&amp;quot;` is present in the font-family of the style attribute which is present in span tag, Hence it would be removed.

Thanks

bansalnitish committed:
Replaced the comment as well

seanlip commented on core/domain/html_cleaner.py:
```diff
@@ -182,7 +182,15 @@ def unescape_html(escaped_html_data):
+    # ensures that all such extra quotes are replaced with
+    # \&amp;quot; in the html string.
+    unescaped_html_data = escaped_html_data.replace(
+        '&amp;quot;', '\\&amp;quot;')
```
> The only way is that we replace &amp;quot; only if it is after font-family attribute in style tags.

If you are sure that this happens only in font-family then include font-family in the check. The more specific the better.

Also I have a question about the current code. What if the data already contains `\\&amp;quot;`?

bansalnitish replied:
If data contains `\\&amp;quot;` it is rendered as `\\&quot;` in the string before unescaping. The quotes after font-family in data are as `&amp;amp;quot;`, so they are rendered as `&amp;quot;` in string before unescaping. So, data will not contain `\\&amp;quot;`.

bansalnitish replied:
The other thing we can do is that `\\` only if they are not present.

seanlip replied:
That would be better.

bansalnitish replied:
What do you suggest here @seanlip @AllanYangZhou ?

seanlip replied:
I thought I just made a suggestion? The stricter the check, the better.

seanlip replied:
Alternatively you can do, before this line, something like: "if string_we're_not_supposed_to_find in data: then error"

bansalnitish replied:
We have two options to make it strict:
- Check for font-family before `&amp;quot;`
- Replace `&amp;quot;` with `\&amp;quot;` only if `\` is missing

I feel it is better to go with second since it will fix this mistake if it is anywhere else as well?

seanlip replied:
Do both?

seanlip replied:
(Though if you have to pick one I lean towards the first. Reasoning: if it was anywhere else, it would have shown up already, right?

At this point I think it's more important to prevent breaking other stuff.)

bansalnitish replied:
For the first check, I need to find the font-family attribute and then look for opening closing `&amp;quot;`. Is there no way except regex for this? 

It's a bit complicated to write regex for the same since there can be multiple pairs of `&amp;quot;`.
```
font-family:&amp;quot;Times New Roman&amp;quot;;\nmso-fareast-font-family:&amp;quot;Times New Roman&amp;quot;
```
Here there are two pairs of `&amp;quot;` and I want to fetch them.

seanlip replied:
I don't think I understand the problem. Why not search for `font-family:&amp;quot;` and replace that string?

bansalnitish replied:
There is an `&amp;quot` in the end as well. 
`font-family:&amp;quot;Times New Roman&amp;quot;`

seanlip replied:
Oh, I see, it's both the start and end?

Should be able to do something like `font-family:&amp;quot;[...]+&amp;quot;`, right? See https://docs.python.org/2/library/re.html#re.sub too.

seanlip replied:
(you'll need to figure out what goes in "...")

bansalnitish committed:
Improved comment and corrected empty content-with-value in collapsible tags

bansalnitish committed:
Made replacement of &amp;quot; more strict

bansalnitish committed:
Update migration according to error cases, add test cases for the same

bansalnitish commented:
Hi @AllanYangZhou @seanlip I am stuck with an issue and need your help here:

From the output which is obtained from dump, I have fixed error cases except one which is as follows:

This is the content-with-value of collapsible tag before unescaping:

```
&quot;&lt;p&gt;A descendant selector can be used to select tags only if they are children of another 
tag.\u00a0&lt;/p&gt;&lt;p&gt;For example if in HTML you 
have;&lt;/p&gt;&lt;blockquote&gt;&lt;p&gt;&lt;b&gt;&lt;ul&gt;\u00a0&lt;/ul&gt;&lt;/b&gt;&lt;/p&gt;&lt;p&gt;&l
t;b&gt;&lt;li&gt; First Bullet &lt;/li&gt;\u00a0&lt;/b&gt;&lt;/p&gt;&lt;p&gt;&lt;b&gt;&lt;li&gt; Second Bullet 
&lt;/li&gt;\u00a0&lt;/b&gt;&lt;/p&gt;&lt;p&gt;&lt;b&gt;&lt;li&gt; Third Bullet 
&lt;/li&gt;&lt;/b&gt;&lt;/p&gt;&lt;p&gt;&lt;b&gt;&lt;/b&gt;&lt;/p&gt;&lt;/blockquote&gt;&lt;p&gt;The CSS for 
this HTML to change only the &lt;li&gt; for &lt;ul&gt; will 
be;&lt;/ul&gt;&lt;/li&gt;&lt;/p&gt;&lt;blockquote&gt;&lt;p&gt;&lt;b&gt;ul\u00a0li{&lt;/b&gt;&lt;/p&gt;&lt;p&gt;
&lt;b&gt;font-size:24px;&lt;/b&gt;&lt;/p&gt;&lt;p&gt;&lt;b&gt;}&lt;/b&gt;&lt;/p&gt;&lt;/blockquote&gt;&quot;
```

After unescaping this changes to:
```
"<p>A descendant selector can be used to select tags only if they are children of another tag.\u00a0</p>
<p>For example if in HTML you have;</p><blockquote><p><b><ul>\u00a0</ul></b></p><p><b><li> 
First Bullet </li>\u00a0</b></p><p><b><li> Second Bullet </li>\u00a0</b></p><p><b><li> Third Bullet 
</li></b></p><p><b></b></p></blockquote><p>The CSS for this HTML to change only the <li> for <ul> 
will be;</ul></li></p><blockquote><p><b>ul\u00a0li{</b></p><p><b>font-size:24px;</b></p><p><b>}
</b></p></blockquote>"
```

Some of the part here is not a part of html but a part of text, this changes to tags after unescaping and that creates issue with validation as well as migration i.e. BeautifulSoup treats them as normal tags and performs migration as well as corrects any malformed content ( In other case <head> is present as part of text and there is no closing tag but BeautifulSoup automatically closes this and also a valiation error is prouduced.)

Can you please guide me as how should I handle this case?

bansalnitish committed:
Fixed issues with escaping and unescaping HTML data

bansalnitish commented:
We (me and @AllanYangZhou ) had a long discussion regarding the above issue and we figured out that this must be due to the different order of replacement in escaping/unescaping html strings. I changed the order to match the front end order and this has also fixed the font-family issue (no extra stuff is required for it now).
The order now is similar to this:
https://github.com/oppia/oppia/blob/71c5b08ab2c9cfd611b91814cd3c8491189cf8b2/core/templates/dev/head/services/HtmlEscaperService.js#L39

AllanYangZhou commented on core/domain/html_cleaner.py:
```diff
@@ -498,6 +504,26 @@ def convert_to_ckeditor(html_data):
+            li.parent.unwrap()
+
+    # Ensure that the children of ol/ul are li/pre.
+    list_tags = ['ol', 'ul']
```
Define this as a constant `LIST_TAGS`

bansalnitish replied:
Done!

AllanYangZhou commented on core/domain/html_cleaner.py:
```diff
@@ -506,11 +532,14 @@ def convert_to_ckeditor(html_data):
             p.wrap(soup.new_tag('li'))
 
-    # Replaces <p><br><p> with <p>&nbsp;</p>.
+    # Replaces <p><br><p> with <p>&nbsp;</p> and <pre>...<br>...</pre>
```
Minor nit: second `<p>` should be `</p>`

bansalnitish replied:
Done!

AllanYangZhou reviewed (approved):
Two minor notes, once addressed LGTM

bansalnitish committed:
Fixed minor issues

seanlip commented:
> We (me and @AllanYangZhou ) had a long discussion regarding the above issue and we figured out that this must be due to the different order of replacement in escaping/unescaping html strings. I changed the order to match the front end order and this has also fixed the font-family issue (no extra stuff is required for it now).

This is absolutely the correct fix and I'm somewhat surprised it wasn't caught earlier. The transformation sequence should follow exactly what happens in the frontend if you want things to match up.

Thanks for finding and fixing it!

bansalnitish commented:
Thanks @seanlip @AllanYangZhou !
